### PR TITLE
gzip autodetection

### PR DESCRIPTION
**Pull request by sethp**
The cloud-init project uses a similar technique to natively support gzipped user-data. The alternative path would be to use a script that decompresses the user data and then calls into cloudinit, but this seemed simple enough to bring into cloudinit itself.

And, unrelated, but in the same commit – without the change to `test`, I found that the top-level tests in coreos-cloudinit_test.go were not running.

### CODE REVIEW

**Comment by sethp** [comment]
@crawford it looks like I missed the great issue migration out of this project, but I think this addresses coreos/bugs#741 – would you be able to give it a look?

**Comment by crawford** [comment]
@sethp I haven't had a chance to look at this yet. Rest assured I haven't forgotten about it!

File: coreos-cloudinit.go [hunk 1]
```diff
@@ -392,3 +401,19 @@ func runScript(script config.Script, env *initialize.Environment) error {
 	}
 	return err
 }
+
+// Via http://cloudinit.readthedocs.org/en/latest/topics/format.html –
+// Content found to be gzip compressed will be uncompressed. The uncompressed data will
+// then be used as if it were not compressed. This is typically is useful because
+// user-data is limited to ~16384 bytes
+func checkForGzip(userdataBytes []byte) ([]byte, error) {
+	if len(userdataBytes) > 1 && 0x1f8b == binary.BigEndian.Uint16(userdataBytes[:2]) {
```
Comment (crawford) [inline coreos-cloudinit.go]: Use `bytes.HasPrefix()` instead.

File: coreos-cloudinit.go [hunk 2]
```diff
@@ -392,3 +401,19 @@ func runScript(script config.Script, env *initialize.Environment) error {
 	}
 	return err
 }
+
+// Via http://cloudinit.readthedocs.org/en/latest/topics/format.html –
+// Content found to be gzip compressed will be uncompressed. The uncompressed data will
+// then be used as if it were not compressed. This is typically is useful because
+// user-data is limited to ~16384 bytes
+func checkForGzip(userdataBytes []byte) ([]byte, error) {
```
Comment (crawford) [inline coreos-cloudinit.go]: Maybe rename to `decompressIfGzip`.

File: coreos-cloudinit.go [hunk 3]
```diff
@@ -392,3 +401,19 @@ func runScript(script config.Script, env *initialize.Environment) error {
 	}
 	return err
 }
+
+// Via http://cloudinit.readthedocs.org/en/latest/topics/format.html –
```
Comment (crawford) [inline coreos-cloudinit.go]: This comment isn't extremely relevant. I would drop it.

File: test [hunk 4]
```diff
@@ -21,6 +21,7 @@ SRC="
 	network
 	pkg
 	system
+	.
```
Comment (crawford) [inline test]: Can you break this into a separate commit?

**Comment by sethp** [comment]
Thank you for the feedback! I think I've addressed each point, would you let me know if you're happy with it?

**Comment by crawford** [comment]
Can you also prefix your commit messages with a scope (e.g. "test: " and "userdata: ")?

**Comment by sethp** [comment]
Ah, sorry – I think my messages match the format in CONTRIBUTING now. Please let me know if there's anything else I missed!

**Comment by crawford** [comment]
This looks good to me. @jonboulle, did you have any concerns?

**Comment by jonboulle** [comment]
LGTM
